File: src/test/java/dev/jlo/gamemodes/domain/common/MatchRosterTest.java
package dev.jlo.gamemodes.domain.common;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.nio.charset.StandardCharsets;
import java.util.UUID;
import java.util.stream.IntStream;
import org.junit.jupiter.api.Test;

class MatchRosterTest {
    @Test
    void assignmentKeepsTeamsBalancedAndRespectsQuorum() {
        MatchRoster roster = new MatchRoster(4, Quorum.minimumPlayers(4, 0.5));
        var players = IntStream.rangeClosed(1, 6)
                .mapToObj(it -> UUID.nameUUIDFromBytes(("player-" + it).getBytes(StandardCharsets.UTF_8)))
                .toList();

        for (UUID player : players) {
            roster.join(player);
        }

        assertEquals(3, roster.players(Team.A).size());
        assertEquals(3, roster.players(Team.B).size());
        assertEquals(true, roster.hasQuorum());
    }

    @Test
    void teamCapacityAndDuplicateJoinsAreRejected() {
        MatchRoster roster = new MatchRoster(1, Quorum.fixed(1));
        UUID first = UUID.randomUUID();
        UUID second = UUID.randomUUID();

        roster.join(first);
        assertThrows(IllegalStateException.class, () -> roster.join(first));
        roster.join(second);
        assertThrows(IllegalStateException.class, () -> roster.join(UUID.randomUUID()));
    }

    @Test
    void quorumRoundsFractionalCapacityUp() {
        assertEquals(2, Quorum.minimumPlayers(3, 0.5).getRequiredPlayers());
        assertEquals(1, Quorum.minimumPlayers(3, 0.1).getRequiredPlayers());
    }
}

File: src/test/kotlin/dev/jlo/gamemodes/domain/common/MatchRosterTest.kt
package dev.jlo.gamemodes.domain.common

import java.util.UUID
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class MatchRosterTest {
    @Test
    fun `assignment keeps teams balanced and respects quorum`() {
        val roster = MatchRoster(teamCapacity = 4, quorum = Quorum.minimumPlayers(4, 0.5))
        val players = (1..6).map { UUID.nameUUIDFromBytes("player-$it".toByteArray()) }

        players.forEach(roster::join)

        assertEquals(3, roster.players(Team.A).size)
        assertEquals(3, roster.players(Team.B).size)
        assertEquals(true, roster.hasQuorum())
    }

    @Test
    fun `team capacity and duplicate joins are rejected`() {
        val roster = MatchRoster(teamCapacity = 1, quorum = Quorum.fixed(1))
        val first = UUID.randomUUID()
        val second = UUID.randomUUID()

        roster.join(first)
        assertFailsWith<IllegalStateException> { roster.join(first) }
        roster.join(second)
        assertFailsWith<IllegalStateException> { roster.join(UUID.randomUUID()) }
    }

    @Test
    fun `quorum rounds fractional capacity up`() {
        assertEquals(2, Quorum.minimumPlayers(3, 0.5).requiredPlayers)
        assertEquals(1, Quorum.minimumPlayers(3, 0.1).requiredPlayers)
    }
}
